### PR TITLE
doc: manual: performance tips: warn against suboptimal type asserts

### PR DESCRIPTION
**Pull request by nsajko**
Add a subsection to the Performance tips.

Updates #38274

### CODE REVIEW

**Comment by mikmoore** [comment]
In the examples given, this suggestion is redundant with input type specialization behavior. I guess this could matter if the function was inlined into a type-unstable context, but that nuance is not explained here. And it would seem that ["Annotate values taken from untyped locations"](https://docs.julialang.org/en/v1/manual/performance-tips/#Annotate-values-taken-from-untyped-locations) would mitigate this issue in most practical contexts, anyway.

We already have users C-style concrete-annotating every function argument while thinking they're doing themselves and others a favor. I'd rather they not get inspired (and encouraged... if only through their misinterpretation) to litter function bodies with `(x isa Integer) || error("not integer")` when the compiler usually knows the concrete type already.

I think this example should use a concrete type like `Int`. Abstract type annotations seldom do much to remove the impact of poor inference so I'd rather not imply that they are ever part of "performance tips compliant" code.

---

Given the complex understanding of code context and the intersection of concepts necessary to even appreciate the tiny slice of situations where this would actually apply (and to understand the overwhelming majority of cases where it does *not*), I think I would prefer not to put this into the performance tips. I would rather field the rare resulting issue on Discourse, where someone can hopefully point to (or rediscover) #38274. As a performance tip, I think it would mostly be misunderstood and applied when inapplicable. I think it risks more harm than good. 

It appears undisputed that the linked issue is something that "should" get fixed, so this performance tip might expire anyway. If the issue hadn't been open for four years already, I would fight that point harder.

Or do I grossly underestimate or misunderstand how often this is relevant, even after "Annotate values taken from untyped locations" is followed?

**Comment by nsajko** [comment]
I made improvements thanks you your message @mikmoore. Can you reevaluate, I think this is worthwhile now.

The added subsection now resides immediately below "Annotate values taken from untyped locations". And the example functions in the added subsection now use `Vector{Any}` ("untyped location").

> It appears undisputed that the linked issue is something that "should" get fixed, so this performance tip might expire anyway. If the issue hadn't been open for four years already, I would fight that point harder.

Well, yeah, it's been open for a long time, and the comments seem to say that fixing the issue would be difficult. A PR was even attempted at one point.

**Comment by JeffBezanson** [comment]
This is a good PR, but this issue can be fixed fairly easily by merging #40880 so we might want to do that instead.

**Comment by fingolfin** [comment]
@JeffBezanson that has been superseded by PR #55229 so hopefully that can be merged soonish?

File: doc/src/manual/performance-tips.md [hunk 1]
```diff
@@ -577,6 +577,45 @@ c = (b + 1.0f0)::Complex{T}
 does not hinder performance (but does not help either) since the compiler can determine the type of `c`
 at the time `k` is compiled.
 
+### Beware of suboptimal type assertion patterns
+
+Due to [limitations](https://github.com/JuliaLang/julia/issues/38274) in current versions
+of Julia, some [type assertion](@ref "Type Declarations") patterns do not help refine type
+inference, even though, intuitively, they should. An example:
```
Comment (adienes) [inline doc/src/manual/performance-tips.md:584]:
```suggestion
inference. An example:
```

IMO documentation should consist of positive statements and omit normative statements.

**Pull request closed without merging**